kernel/src/attest.rs

```
@@ -0,0 +1,262 @@
+impl From<Tee> for AttestationDriver<'_> {
+    fn from(tee: Tee) -> Self {
+        let sp = SerialPort::new(&DEFAULT_IO_DRIVER, 0x3e8); // COM3
+        sp.init();
+
+        Self { sp, tee, key: None }
+    }
+}
+
+impl AttestationDriver<'_> {
+    /// Attest SVSM's launch state by communicating with the attestation proxy.
+    pub fn attest(&mut self) -> Vec<u8> {
+        let negotiation = self.negotiation();
+
+        self.attestation(negotiation)
+    }
+
+    /// Send a negotiation request to the proxy. Proxy should reply with Negotiation parameters
+    /// that should be included in attestation evidence (e.g. through SEV-SNP's REPORT_DATA
+    /// mechanism).
+    fn negotiation(&mut self) -> NegotiationResponse {
+        let request = NegotiationRequest {
+            version: "0.1.0".to_string(), // Only version supported at present.
+            tee: self.tee,
+        };
+
+        self.write(request);
+
+        let response: NegotiationResponse = {
+            let payload = self.read();
+
+            serde_json::from_slice(&payload).unwrap()
+        };
+
+        response
+    }
+
+    /// Send an attestation request to the proxy. Proxy should reply with attestation response
+    /// containing the status (success/fail) and an optional secret returned from the server upon
+    /// successful attestation.
+    fn attestation(&mut self, negotiation: NegotiationResponse) -> Vec<u8> {
+        // Gen
+        self.tee_key_generate(&negotiation);
+        let evidence = self.evidence(negotiation);
+
+        let request = AttestationRequest {
+            evidence: BASE64_STANDARD.encode(evidence),
+            key: AttestationKey::from(&self.key.clone().unwrap()),
+        };
+
+        self.write(request);
+
+        let response: AttestationResponse = {
+            let payload = self.read();
+
+            serde_json::from_slice(&payload).unwrap()
+        };
+
+        if !response.success {
+            panic!("attestation failed");
+        }
```

Can we return an error and have the panic in the main function (e.g. when calling `driver.attest()`). Feel free to introduce an `AttestationError` and a new variant for `SvsmError`

---

Sure. What about the rest of the `unwrap()`s and `panic()`s in the file? Should they be replaced with returns of `AttestationError` as well?

---

I'd say so, yes. The distinction between what to return and what to panic on immediately is subtle, but in general I'd favor panicking as high on the stack as possible. Unless it's a hard bug condition, I'd just bubble it up.

---

Addressed in https://github.com/coconut-svsm/svsm/pull/528/commits/8f5b6774cb0989943f1bd31dabda34138a6d2ae1, can you please re-review?

---

Polite ping @00xc, can I resolve this conversation?

---

Resolving this for now. If desired we can review this at a later point?

---

Sorry, missed this! LGTM